mm2src/coins/eth.rs

```diff
@@ -1468,48 +1435,75 @@ impl WatcherOps for EthCoin {
                 )));
             }
 
+            if tx_from_rpc.to != Some(expected_swap_contract_address) {
+                if tx_from_rpc.to == fallback_swap_contract {
+                    expected_swap_contract_address = fallback_swap_contract.ok_or_else(|| {
+                        ValidatePaymentError::InternalError("Fallback swap contract address not found".to_string())
+                    })?;
+                } else {
+                    return MmError::err(ValidatePaymentError::WrongPaymentTx(format!(
+                        "Payment tx {:?} was sent to wrong address, expected {:?}",
+                        tx_from_rpc, expected_swap_contract_address,
+                    )));
+                }
+            }
+
+            let status = selfi
+                .payment_status(expected_swap_contract_address, Token::FixedBytes(swap_id.clone()))
+                .compat()
+                .await
+                .map_to_mm(ValidatePaymentError::Transport)?;
+            if status != PAYMENT_STATE_SENT.into() {
+                return MmError::err(ValidatePaymentError::UnexpectedPaymentState(format!(
+                    "Payment state is not PAYMENT_STATE_SENT, got {}",
+                    status
+                )));
+            }
+
             match &selfi.coin_type {
                 EthCoinType::Eth => {
-                    if tx_from_rpc.to != Some(expected_swap_contract_address) {
-                        return MmError::err(ValidatePaymentError::WrongPaymentTx(format!(
-                            "Payment tx {:?} was sent to wrong address, expected {:?}",
-                            tx_from_rpc, expected_swap_contract_address,
-                        )));
-                    }
-
                     let function = SWAP_CONTRACT
                         .function("ethPayment")
                         .map_to_mm(|err| ValidatePaymentError::InternalError(err.to_string()))?;
                     let decoded = function
                         .decode_input(&tx_from_rpc.input.0)
                         .map_to_mm(|err| ValidatePaymentError::TxDeserializationError(err.to_string()))?;
-                    if decoded[0] != Token::FixedBytes(swap_id.clone()) {
+
+                    let swap_id_input = get_function_input_data(&decoded, function, 0)
+                        .map_to_mm(ValidatePaymentError::TxDeserializationError)?;
```

Comment:
It's not critical. In my opinion, the best way to handle each token step by step is to use `Iterator::next`.
But only if the tokens are handled one after another.
For example,
https://github.com/KomodoPlatform/atomicDEX-API/blob/2e720274af8173b42a50fa54bbf36517096bbb51/mm2src/coins/qrc20/swap.rs#L681-L696

Reply:
We have too many decoded index accesses in the eth module, it would take hundreds of lines if we handle them all like this. Also in some cases not all sequential array items are required as you said, and I don't like the idea of using next() to skip unused items. I decided to use a helper function to simplify this as much as possible, and I think it'd be better if we use this and maybe improve it.

Reply:
If we need more detailed error messages, we could get e.g. the parameter names from the `Function` input, or we could use a different helper for each contract function (e.g. erc20Payment) and write everything manually. But in general I'd prefer a solution that does all of this once, and not every time we use them. Maybe we can open an issue for this, there's still lots of direct index accesses in the eth module that I didn't fix.

Reply:
I meant that if you need to parse 1st, 2nd, 3rd, ... tokens, you can use `Iterator::next`.
> The best way to handle each token step by step is to use `Iterator::next`. But only if the tokens are handled one after another.

In other cases, I like your solution to use `get_function_input_data`.
Anyway, the way you've chosen definitely can be